topic/machine-learning/llm-langchain/cratedb_rag_customer_support.ipynb

```diff
@@ -0,0 +1,445 @@
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "</div>"
+      ],
+      "text/plain": [
+       "                                                text   \n",
+       "0  tweet_id: 2301\\nauthor_id: 116231\\ninbound: Tr...  \\\n",
+       "1  tweet_id: 11879\\nauthor_id: MicrosoftHelps\\nin...   \n",
+       "2  tweet_id: 11881\\nauthor_id: MicrosoftHelps\\nin...   \n",
+       "3  tweet_id: 11890\\nauthor_id: 118332\\ninbound: T...   \n",
+       "4  tweet_id: 11912\\nauthor_id: MicrosoftHelps\\nin...   \n",
+       "\n",
+       "                                           embedding  \n",
+       "0  [-0.03669742581690742, -0.013565617003293586, ...  \n",
+       "1  [-0.015454164058839013, 0.0032340502581370404,...  \n",
+       "2  [-0.005936504790842901, 0.01942733669848252, 0...  \n",
+       "3  [-0.01177901347977142, 0.005725434705161641, -...  \n",
+       "4  [-0.022950152341946858, 0.004767860370434741, ...  "
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "df.to_sql(name='text_data', con=engine, if_exists='append', index=False)\n",
+    "df.head(5)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "GkAPZ55RZQ09"
+   },
+   "source": [
+    "#Ask question\n",
+    "Let's define our question and create an embedding using OpenAI embedding model:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "InhR73isZJCB"
+   },
+   "outputs": [],
+   "source": [
+    "my_question = \"How to update shipping address on existing order in Microsoft Store?\"\n",
+    "query_embedding = embeddings.embed_query(my_question)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "6XnNZHI6ajaS"
+   },
+   "source": [
+    "#Find relevant context using similarity search\n",
+    "\n",
+    "The `knn_match (search_vector, query_vector, k) `function in CrateDB performs an approximate k-nearest neighbors (KNN) search within a dataset. KNN search involves finding the k data points that are most similar to a given query data point. We find the most similar vectors to our query vector using knn search capability in CrateDB:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "VjLeMkwMagOf",
+    "outputId": "2c92d6fc-22aa-4914-b58c-bbd3928108e1"
+   },
+   "outputs": [],
+   "source": [
+    "knn_query = text(\"\"\"SELECT text FROM text_data\n",
+    "            WHERE knn_match(embedding, {0}, 2)\"\"\".format(query_embedding))\n",
+    "documents=[]\n",
+    "\n",
+    "with engine.connect() as con:\n",
+    "    results = con.execute(knn_query)\n",
+    "    for record in results:\n",
+    "        documents.append(record[0])\n",
+    "\n",
+    "print(documents)\n"
```

Comment:
a) What does `query_embedding` interpolate here, i.e. of which Python type is it? b) I am gravitating towards submitting it as a real query parameter instead of interpolating it into the query expression.

Reply:
In Python terms it is `List[float]`. This example is significantly less important than the one with the LangChain

Reply:
All right. Then, this code works only by accident, and because the printed serialization of a Python list apparently is the same syntax to feed it into the SQL statement. I don't know if I should be happy about it.